### PR TITLE
Tutorial should allow user to set location by sharing it

### PR DESCRIPTION
**Pull request by arthurshir**
## Changes
* Amend tutorial/eventIdentifier to identify events by message rather than utterance
* Implement telegram location event identification

## Testing Done
* Implemented unit tests

### CODE REVIEW

File: botfunctions/timezonecommand.js [hunk 1]
```diff
@@ -199,6 +199,20 @@ function isValidUtterance(utterance) {
     return utils.isValidUtterance(utterance, isValidCommandInput, COMMAND_REGEX, constants.COMMAND_TYPE.SLASH_OR_TEXT_COMMAND);
 }
 
+function isValidLocation(location) {
+    return new Promise((resolve) => {
+        if (!location.latitude || !location.longitude) {
+            resolve(false);
```
Comment (bubakazouba) [inline botfunctions/timezonecommand.js:205]: All of the logic you added here to make sure the function doesn't crash is good to have inside `_convertCoordinatesToTimezone`, we want it to always run not only for `isValidLocation`.

Can you move this argument checking inside, along with line 207 where you parse the lat, long to Number.

+ the part where you .catch and resolve(false)

After you move all of this you should be left with a simple `_convertCoordinatesToTimezone(location.lat, location.long)`

File: tst/botfunctions/timezonecommand.test.js [hunk 2]
```diff
@@ -1,13 +1,29 @@
...
+jest.mock('@google/maps', () => ({
+    createClient: () => ({timezone: mockTimezone})
+}));
+
 const timezonecommand = require("../../botfunctions/timezonecommand.js"),
     constants = require("../../botfunctions/constants.js");
 
+const GOOGLE_MAPS_SUCCESS_RESPONSE = [null, {json: {timeZoneId: 1}}];
+const GOOGLE_MAPS_ERROR_RESPONSE = ['error', null];
+const GENERIC_LOCATION_INPUT = {latitude: 1, longitude: 1};
```
Comment (bubakazouba) [inline tst/botfunctions/timezonecommand.test.js:11]: NIt can you make this a float, I'd be worried about someone changing `Number` to `parseInt`, unit tests would catch this

**Review by bubakazouba — COMMENTED** [review]
did you try sending a photo and making sure the bot doesn't crash?

File: botfunctions/timezonecommand.js (hunk 1 above)
Comment (arthurshir) [reply botfunctions/timezonecommand.js:205]: Oh nice good catch, forgot to consider changing the existing function.

**Comment by arthurshir** [comment]
> did you try sending a photo and making sure the bot doesn't crash?

Verified just now

File: botfunctions/timezonecommand.js (hunk 1 above)
Comment (bubakazouba) [reply botfunctions/timezonecommand.js:205]: or even better `isValidLocation = _convertCoordinatesToTimezone` if you make the interface match

Comment (arthurshir) [reply botfunctions/timezonecommand.js:205]: If it sounds good with you, I'm going to keep that `_convertCoordinatesToTimezone` private

File: tutorial/eventIdentifier.js [hunk 3]
```diff
@@ -26,6 +20,27 @@ function identifyEventFromUtterance(utterance) {
...
 
+/**
+ * Identifies the tutorial event a message matches with.
+ * @param {ctx.message} message - message object of form {text: String} or {location: {latitude: Number, longitude: Number}}
+ * @returns {Promise.<String>} - Promise resolving to a String representing the tutorial event type the message matches with.
+ */
+function identifyEventFromMessage(message) {
+    return new Promise((resolve) => {
+        if (!!message.text) {
+            resolve(_identifyEventFromUtterance(message.text));
```
Comment (bubakazouba) [inline tutorial/eventIdentifier.js:31]: Nit: just return here and remove the else for less nesting

File: tutorial/eventIdentifier.js [hunk 4]
```diff
@@ -26,6 +20,27 @@ function identifyEventFromUtterance(utterance) {
...
+        } else {
+            if (!!message.location) {
+                timezoneCommand.isValidLocation(message.location).then(isValid => {
```
Comment (bubakazouba) [inline tutorial/eventIdentifier.js:34]: return here and remove the next else

**Review by bubakazouba — APPROVED** [review]

**Pull request merged**